heavyball/chainable.py

```diff
@@ -204,6 +204,459 @@ def update_by_laprop(group, update, grad, param, exp_avg, exp_avg_sq):
+    next_update = state.get('next_hessian_update', 1)
+
+    # Check if it's time to update Hessian
+    if state['hessian_step'] >= next_update:
+        state['next_hessian_update'] = state['hessian_step'] + k
+
+        # Compute Hessian estimate based on the type
+        if group.get('sophia_type', 'h') == 'h':
+            # Hutchinson's estimator
+            h = _hutchinson_hessian(state, param, grad)
+        else:
+            # Gauss-Newton-Bartlett estimator
+            h = grad * grad  # Simplified GNB estimator
+
+        # Update EMA of diagonal Hessian
+        if 'diag_hessian' in state:
+            state['diag_hessian'].mul_(beta2).add_(h, alpha=1-beta2)
+        else:
+            state['diag_hessian'] = h.clone()
+
+        # Ensure positive values for numerical stability
+        state['diag_hessian'].clamp_(min=1e-6)
+
+@general_guard(["diag_hessian", "hessian_step", "next_hessian_update"], init_fn=_init_hessian_estimator, skip_first=False)
+@no_state
+def estimate_hessian_g(state, group, update, grad, param):
+    """Estimate diagonal Hessian using Gauss-Newton-Bartlett method."""
+    k = group.get('sophia_update_freq', 10)
+    beta2 = utils.get_beta2(group)
+    state['hessian_step'] += 1
+
+    # Check if it's time to update the Hessian estimate
+    if state['hessian_step'] >= state['next_hessian_update']:
+        state['next_hessian_update'] = state['hessian_step'] + k
+
+        # For GNB we can use a simplified approximation based on squared gradients
+        h = grad * grad  # Simplified GNB estimator
+
+        # Update EMA of diagonal Hessian
+        if 'diag_hessian' in state:
+            state['diag_hessian'].mul_(beta2).add_(h.abs(), alpha=1-beta2)
+        else:
+            state['diag_hessian'] = h.abs().clone()
+
+        # Ensure positive values for numerical stability
+        state['diag_hessian'].clamp_(min=1e-6)
+
+    return update
+
+@zero_guard("exp_avg", "diag_hessian")
+@no_state
+def scale_by_sophia(group, update, grad, param, exp_avg, diag_hessian):
+    """
+    Sophia optimizer update rule. Divides the EMA of gradients by the EMA of the diagonal Hessian,
+    and then applies element-wise clipping.
+    """
+    # Update momentum using beta1
+    beta1 = utils.get_beta1(group)
+    utils.scale_by_exp_avg_(exp_avg, update, utils.beta_debias(beta1, group["step"]))
+
+    # Get clipping threshold
+    gamma = group.get('sophia_gamma', 0.01)
+    eps = group.get('eps', 1e-12)
+
+    # For each parameter
+    for u, m, h in zip(update, exp_avg, diag_hessian):
+        # Computation for each parameters : mt / max(γ * ht, ε)
+        m32 = utils.promote(m)
+        h32 = utils.promote(h)
+
+        # Compute denominator with safeguard
+        denom = torch.maximum(gamma * h32, torch.tensor(eps, device=h32.device, dtype=h32.dtype))
+
+        # Scale and clip the update
+        scaled = m32 / denom
+        torch.clamp_(scaled, min=-1.0, max=1.0, out=scaled)
```

Comment:
torch.Tensor.clamp_(x) is already an inplace-op, no need to specify `out`. also, `scaled.clamp_(min=-1.0, max=1.0)` may be simpler :)